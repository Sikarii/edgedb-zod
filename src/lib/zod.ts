import { $ } from "edgedb";

const zodType = (name: string, ...args: unknown[]) => `${name}(${args})`;

const zodEnum = (values: unknown[]) => {
  const str = values.join(", ");
  return zodType("enum", `[${str}]`);
};

export const scalarToZod = (type: $.introspect.ScalarType) => {
  if (type.enum_values) {
    const values = type.enum_values.map((v) => `"${v}"`);

    return [
      zodEnum(values),
    ];
  }

  switch (type.name) {
    case "std::str":
      return [
        zodType("string"),
      ];

    case "std::bool":
      return [
        zodType("boolean"),
      ];

    case "std::json":
      return [
        zodType("unknown"),
      ];

    case "std::uuid":
      return [
        zodType("string"),
        zodType("uuid"),
      ];

    case "std::int16":
      return [
        zodType("number"),
        zodType("int"),
        zodType("min", 0),
        zodType("max", 65535),
      ];

    case "std::int32":
      return [
        zodType("number"),
        zodType("int"),
        zodType("min", 0),
        zodType("max", 2147483647),
      ];

    case "std::int64":
      return [
        zodType("number"),
        zodType("int"),
        zodType("min", 0),
      ];

    case "std::bigint":
      return [
        zodType("bigint"),
      ];

    case "std::datetime":
      return [
        zodType("date"),
      ];

    case "std::float32":
      return [
        zodType("number"),
        zodType("min", 0),
        zodType("max", 3.40282347E+38),
      ];

    case "std::float64":
      return [
        zodType("number"),
        zodType("min", 0),
        zodType("max", 1.7976931348623157E+308),
      ];

    // This is not a type in EdgeDB, the JS driver has this
    case "std::number":
      return [
        zodType("number"),
      ];

    default:
      // TODO: Null might make more sense
      return [
        zodType("never"),
      ];
  }
};
