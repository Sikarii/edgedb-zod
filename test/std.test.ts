import { z } from "zod";
import { it, describe, expect } from "vitest";

const std = import("../dbschema/edgedb-zod/modules/std");

describe("std::int16", async () => {
  const schema = (await std).int16Schema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodNumber);
  });

  it("it does the correct checks", () => {
    expect(schema._def.checks).toMatchSnapshot();
  });
});

describe("std::int32", async () => {
  const schema = (await std).int32Schema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodNumber);
  });

  it("it does the correct checks", () => {
    expect(schema._def.checks).toMatchSnapshot();
  });
});

describe("std::int64", async () => {
  const schema = (await std).int64Schema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodNumber);
  });

  it("it does the correct checks", () => {
    expect(schema._def.checks).toMatchSnapshot();
  });
});

describe("std::str", async () => {
  const schema = (await std).strSchema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodString);
  });
});

describe("std::bool", async () => {
  const schema = (await std).boolSchema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodBoolean);
  });
});

describe("std::json", async () => {
  const schema = (await std).jsonSchema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodUnknown);
  });
});

describe("std::uuid", async () => {
  const schema = (await std).uuidSchema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodString);
  });

  it("it does the correct checks", () => {
    expect(schema._def.checks).toMatchSnapshot();
  });
});

describe("std::number", async () => {
  const schema = (await std).numberSchema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodNumber);
  });
});

describe("std::datetime", async () => {
  const schema = (await std).datetimeSchema;

  it("is exported", () => {
    expect(schema).toBeDefined();
  });

  it("is the correct instance", () => {
    expect(schema).toBeInstanceOf(z.ZodString);
  });

  it("it does the correct checks", () => {
    expect(schema._def.checks).toMatchSnapshot();
  });
});
