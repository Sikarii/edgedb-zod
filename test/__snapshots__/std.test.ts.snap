// Vitest Snapshot v1

exports[`std::datetime > it does the correct checks 1`] = `
[
  {
    "kind": "regex",
    "regex": /\\^\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}\\(T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\(\\\\\\.\\\\d\\{3\\}\\)\\?\\)\\?Z\\?\\$/,
  },
]
`;

exports[`std::int16 > it does the correct checks 1`] = `
[
  {
    "kind": "int",
    "message": undefined,
  },
  {
    "inclusive": true,
    "kind": "min",
    "message": undefined,
    "value": 0,
  },
  {
    "inclusive": true,
    "kind": "max",
    "message": undefined,
    "value": 65535,
  },
]
`;

exports[`std::int32 > it does the correct checks 1`] = `
[
  {
    "kind": "int",
    "message": undefined,
  },
  {
    "inclusive": true,
    "kind": "min",
    "message": undefined,
    "value": 0,
  },
  {
    "inclusive": true,
    "kind": "max",
    "message": undefined,
    "value": 2147483647,
  },
]
`;

exports[`std::int64 > it does the correct checks 1`] = `
[
  {
    "kind": "int",
    "message": undefined,
  },
  {
    "inclusive": true,
    "kind": "min",
    "message": undefined,
    "value": 0,
  },
]
`;

exports[`std::uuid > it does the correct checks 1`] = `
[
  {
    "kind": "uuid",
  },
]
`;
